lumicks/pylake/detail/bead_cropping.py

```diff
@@ -88,14 +105,178 @@ def find_beads(
+    threshold_percentile : int
+        Percentile below which to drop.
+
+    Returns
+    -------
+    list of int
+       List of the two edge positions in pixels.
+
+    Raises
+    ------
+    ValueError
+        If num_time_frames is set to a value lower than 3.
+    ValueError
+        If the kymograph image is shorter than 3 frames or has incorrect dimensions.
+    """
+    import scipy.ndimage
+    import matplotlib.pyplot as plt
+    from skimage.measure import block_reduce
+
+    if num_time_frames < 3:
+        raise ValueError(
+            f"Time axis needs to be divided in at least 3 sections, provided {num_time_frames}"
+        )
+
+    if kymograph_image.ndim != 2:
+        raise ValueError(
+            f"Kymograph image must be two dimensional. Provided {kymograph_image.ndim} dimensional "
+            "image"
+        )
+
+    if kymograph_image.shape[1] < 3:
+        raise ValueError(
+            "Kymograph is too short to apply this method. This method requires at least 3 scan "
+            f"lines while this kymograph has {kymograph_image.shape[1]} scan lines."
+        )
+
+    half_width = int(bead_diameter_pixels // 2)
+    template_size_pixels = half_width * 2 + 1
+
+    ds_factor = kymograph_image.shape[1] // num_time_frames
+    img = (
+        block_reduce(kymograph_image, (1, ds_factor), func=np.sum)[
+            :, : kymograph_image.shape[1] // ds_factor
+        ]
+        if ds_factor > 1
+        else kymograph_image
+    )
+
+    # Try to guess the background by finding the most prominent mode
+    baseline = _guess_background(img.flatten())
+
+    # Flip dark beads into having positive signal that we can template match on
+    img = np.abs(img - baseline)
+
+    # Pad so that we don't get boundary effects from our filtering
+    padding = np.zeros((half_width * 2, img.shape[1]))
+    padded_img = np.vstack((padding, img, padding))
+
+    # Set up the correlation matrix. For each frame, we apply a sliding window.
+    # for each window we look into the subsequent frames whether we see the window there
+    # using cross correlation.
+    correlation_stack = []
+    for template_origin in range(padded_img.shape[0] - template_size_pixels):
+        correlation_image = []
+
+        # Calculate the successive correlation between frames
+        for t in range(padded_img.shape[1] - 1):
+            template = padded_img[template_origin : template_origin + template_size_pixels, t]
+            normalization_factor = 1.0 + np.std(padded_img[:, t + 1]) * np.std(template)
+
+            if allow_movement:
+                correlation_image.append(
+                    np.max(scipy.signal.correlate(padded_img[:, t + 1], template, mode="same"))
+                    / normalization_factor
+                )
+            else:
+                im = padded_img[template_origin : template_origin + template_size_pixels, t + 1]
+                correlation_image.append(np.sum(im * template) / normalization_factor)
```

Comment:
would it be worth it to try to pull this out in a separate function? the double `for` loop using `append`s could then become a pretty simple nested list comprehension I think. Or do you think this would be a big hit for performance because of significant function calls?

Reply:
Nah, I'm not too worried about perf here. Most of the time is gonna be spent in `correlate`. For `allow_movement=False`, there's a faster option (see last commit), but you get two pretty distinct branches then because of the normalization required. Not sure if this is worth the harder to read code for a function that is pretty fast.